3.processing_features/scripts/2.pycytominer_singlecell_pipelines.py

```diff
@@ -0,0 +1,157 @@
+            plate_info_dictionary["Plate_3"]["platemap_path"]
+        )
+
+# view the dictionary to assess that all info is added correctly
+pprint.pprint(plate_info_dictionary, indent=4)
+
+
+# ## Perform single-cell pycytominer pipeline
+
+# In[4]:
+
+
+for plate, info in plate_info_dictionary.items():
+    print(f"Now performing single-cell pycytominer pipeline for {plate}")
+    output_annotated_file = str(pathlib.Path(f"{output_dir}/{plate}_sc_annotated.parquet"))
+    output_normalized_file = str(pathlib.Path(f"{output_dir}/{plate}_sc_normalized.parquet"))
+    output_feature_select_file = str(pathlib.Path(f"{output_dir}/{plate}_sc_feature_selected.parquet"))
+    output_feature_select_file = str(pathlib.Path(f"{output_dir}/{plate}_sc_feature_selected.parquet"))
+    output_aggregated_file = str(pathlib.Path(f"{output_dir}/{plate}_bulk_camerons_method.parquet"))
+
+    # Load single-cell profiles
+    single_cell_df = pd.read_parquet(info["dest_path"])
+    
+    # Load platemap
+    platemap_df = pd.read_csv(info["platemap_path"])
+    
+    # Step 1: Annotation
+    # add metadata from platemap file to extracted single cell features
+    annotated_df = annotate(
+        profiles=single_cell_df,
+        platemap=platemap_df,
+        join_on=["Metadata_well_position", "Image_Metadata_Well"],
+    )
+
+    # rename site column to avoid any issues with identifying the column as metadata over feature
+    annotated_df = annotated_df.rename(columns={"Image_Metadata_Site": "Metadata_Site"})
+
+    # move metadata well, single cell count, and site to the front of the df (for easy visualization in python)
+    well_column = annotated_df.pop("Metadata_Well")
+    singlecell_column = annotated_df.pop("Metadata_number_of_singlecells")
+    site_column = annotated_df.pop("Metadata_Site")    
+
+    # insert the columns in specific parts of the dataframe
+    annotated_df.insert(2, "Metadata_Well", well_column)
+    annotated_df.insert(3, "Metadata_Site", site_column)
+    annotated_df.insert(4, "Metadata_number_of_singlecells", singlecell_column)
+
+    # save annotated df as parquet file
+    output(
+        df=annotated_df,
+        output_filename=output_annotated_file,
+        output_type="parquet",
+    )
+    
+    # Step 2: Normalization
+    normalized_df = normalize(
+        profiles=output_annotated_file,
+        method="standardize",
+        output_file=output_normalized_file,
+        output_type="parquet",
+    )
+    
+    # Step 3: Feature selection
+    feature_select(
+        output_normalized_file,
+        operation=feature_select_ops,
+        output_file=output_feature_select_file,
+        output_type="parquet"
+    )
+
+    # Step 4: Cameron's method of aggregation
+    feature_select_df = load_profiles(output_feature_select_file)
+    metadata_cols = infer_cp_features(feature_select_df, metadata=True)
+    metadata_cols = [x for x in metadata_cols if all(col not in x for col in cameron_unwanted_aggregate_cols)]
```

Comment:
Based on my understanding, this seems to remove columns that would be variable in each of the wells (e.g., in Well A6 there are 4 different sites so groupby wouldn't work).

It is a bit confusing to me since that means that aggregation is using all the rest of the metadata to groupby. But, since the rest of the metadata is all the same for each Well and Plate metadata (e.g., Well A6 is always one treatment, one siRNA, one concentration, etc.), then wouldn't it be the same as only specifying the strata as ["Metadata_Plate", "Metadata_Well"]?

Reply:
One of the downsides of a long time passing between PR review and my responses, is that I have forgotten the exact need for doing this. 

This is also a symptom of me being a bit lazy and not adding sufficient documentation. 

IIRC, the aggregation function drops metadata columns that aren't specified by `strata`. By including all metadata columns here that we know would have the same effect feature-wise ensures that they are kept in the output dataframe. Now that I type that out, it sounds like exactly the reason I did this... It's not a great pycytominer requirement, and one would think that there should be a function argument to perform this, but alas.

Reply:
I'll add a comment to increase clarity, thanks!

Reply:
added in bc9fce06f6241ddb48fdaacf1892339d3bd1f3c6

Reply:
This does clarify it much more!